pkg/diff/chunk.go

```diff
@@ -17,260 +17,530 @@ import (
-	// first select the index, and number type index first
-	column, err := dbutil.FindSuitableColumnWithIndex(context.Background(), db, Schema, table)
-	if err != nil {
-		return nil, errors.Trace(err)
+// getSplitFields returns fields to split chunks, order by pk, uk, index, columns.
+func getSplitFields(db *sql.DB, schema string, table *model.TableInfo, splitFields []string) ([]*model.ColumnInfo, error) {
+	cols := make([]*model.ColumnInfo, 0, len(table.Columns))
+	colsMap := make(map[string]interface{})
+
+	splitCols := make([]*model.ColumnInfo, 0, 2)
+	for _, splitField := range splitFields {
+		col := dbutil.FindColumnByName(table.Columns, splitField)
+		if col == nil {
+			return nil, errors.NotFoundf("column %s in table %s", splitField, table.Name)
+
+		}
+		splitCols = append(splitCols, col)
 	}
-	if column != nil {
-		return column, nil
+
+	indexColumns := dbutil.FindAllColumnWithIndex(table)
+
+	// user's config had higher priorities
+	for _, col := range append(append(splitCols, indexColumns...), table.Columns...) {
+		if _, ok := colsMap[col.Name.O]; ok {
+			continue
+		}
+
+		colsMap[col.Name.O] = struct{}{}
+		cols = append(cols, col)
 	}
 
-	// use the first column
-	log.Infof("%s.%s don't have index, will use the first column as split field", Schema, table.Name.O)
-	return table.Columns[0], nil
+	return cols, nil
+}
+
+// CheckJob is the struct of job for check
+type CheckJob struct {
+	Schema string
+	Table  string
+	Where  string
+	Args   []string
 }
 
 // GenerateCheckJob generates some CheckJobs.
-func GenerateCheckJob(table *TableInstance, splitField, limits string, chunkSize, sample int, collation string) ([]*CheckJob, error) {
+func GenerateCheckJob(table *TableInstance, splitFields, limits string, chunkSize int, collation string, useTiDBStatsInfo bool) ([]*CheckJob, error) {
 	jobBucket := make([]*CheckJob, 0, 10)
 	var jobCnt int
-	var column *model.ColumnInfo
 	var err error
 
-	if splitField == "" {
-		column, err = findSuitableField(table.Conn, table.Schema, table.info)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-	} else {
-		column = dbutil.FindColumnByName(table.info.Columns, splitField)
-		if column == nil {
-			return nil, errors.NotFoundf("column %s in table %s", splitField, table.Table)
-		}
+	var splitFieldArr []string
+	if len(splitFields) != 0 {
+		splitFieldArr = strings.Split(splitFields, ",")
+	}
+
+	fields, err := getSplitFields(table.Conn, table.Schema, table.info, splitFieldArr)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	chunks, err := getChunksForTable(table, column, chunkSize, sample, limits, collation)
+	chunks, mode, err := getChunksForTable(table, fields, chunkSize, limits, collation, useTiDBStatsInfo)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	if chunks == nil {
 		return nil, nil
 	}
-	log.Debugf("chunks: %+v", chunks)
 
 	jobCnt += len(chunks)
 
-	if collation != "" {
-		collation = fmt.Sprintf(" COLLATE \"%s\"", collation)
-	}
-
 	for {
```

Comment:
`for _, chunk := range chunks {` is more readable?
and seems `jobCnt` is unused.